src/dmd/must_use.d

```diff
@@ -0,0 +1,198 @@
+        }
+    }
+    return false;
+}
+
+/**
+ * Called from a symbol's semantic to check for reserved usage of @mustUse.
+ *
+ * If such usage is found, emits an errror.
+ *
+ * Params:
+ *   sym = symbol to check
+ */
+void checkMustUseReserved(Dsymbol sym)
+{
+    import dmd.errors : error;
+    import dmd.id : Id;
+
+    if (sym.userAttribDecl is null || sym.userAttribDecl.atts is null)
+        return;
+
+    // Can't use foreachUda (and by extension hasMustUseAttribute) while
+    // semantic analysis of `sym` is still in progress
+    // TODO: factor out common code from this function and checkGNUABITag
+    foreach (exp; *sym.userAttribDecl.atts)
+    {
+        if (isMustUseAttribute(exp))
+        {
+            if (sym.isFuncDeclaration())
+            {
+                error(sym.loc, "`@%s` on functions is reserved for future use",
+                    Id.udaMustUse.toChars());
+                sym.errors = true;
+            }
+            else if (sym.isClassDeclaration() || sym.isEnumDeclaration())
+            {
+                error(sym.loc, "`@%s` on `%s` types is reserved for future use",
+                    Id.udaMustUse.toChars(), sym.kind());
+                sym.errors = true;
+            }
+        }
+    }
+}
+
+/**
+ * Returns true if the given expression is an assignment, either simple (a = b)
+ * or compound (a += b, etc).
+ */
+private bool isAssignment(Expression e)
+{
+    import dmd.common.string : startsWith, endsWith;
+
+    if (e.isAssignExp || e.isBinAssignExp)
+        return true;
+    if (auto ce = e.isCallExp())
+    {
+        auto fd = ce.f;
+        auto id = fd ? fd.ident : null;
+        // opXXX are reserved identifiers, so it's ok to match too much here
+        if (id && id.toString().startsWith("op") && id.toString().endsWith("Assign"))
+            return true;
+    }
+    return false;
+}
+
+/**
+ * Returns true if the given expression is an increment (++) or decrement (--).
+ */
+private bool isIncrementOrDecrement(Expression e)
+{
+    import dmd.dtemplate : isExpression;
+    import dmd.globals : Loc;
+    import dmd.id : Id;
+    import dmd.tokens : EXP;
+
+    if (e.op == EXP.plusPlus
+        || e.op == EXP.minusMinus
+        || e.op == EXP.prePlusPlus
+        || e.op == EXP.preMinusMinus)
+        return true;
+    if (auto call = e.isCallExp())
+    {
+        // Check for overloaded preincrement
+        // e.g., a.opUnary!"++"
+        auto fd = call.f;
+        auto id = fd ? fd.ident : null;
+        if (id == Id.opUnary)
+        {
+            auto ti = fd.parent ? fd.parent.isTemplateInstance() : null;
+            auto tiargs = ti ? ti.tiargs : null;
+            if (tiargs && tiargs.length >= 1)
+            {
+                auto argExp = (*tiargs)[0].isExpression();
+                auto op = argExp ? argExp.isStringExp() : null;
+                scope plusPlus = new StringExp(Loc.initial, "++");
+                scope minusMinus = new StringExp(Loc.initial, "--");
```

Comment:
I'm concerned these allocations will become a major performance problem and memory consumption problem, as they are called for a large fraction of all the expression nodes.

Reply:
Changed from `scope` to `static const`.